### PR TITLE
Propagate partition properties for full outer join

### PR DESCRIPTION
**Pull request by rongrong**
For full outer join, if both sides of the equi-join contains all partitioning columns,
the output of the join would be partitioned on coalesce of these columns.

Fixed the issue with incorrectly optimize out remote exchange for aggregation. Added new test to verify aggregation is correct. Also changed existing test to be more robust.

### CODE REVIEW

**Comment by rongrong** [comment]
@kokosing

**Comment by rongrong** [comment]
Reintroduces feature reverted by #12351

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                "FROM (" +
+                "   SELECT coalesce(t.a, s.a) AS a " +
+                "   FROM (VALUES (1), (2), (3)) t(a) " +
+                "   FULL OUTER JOIN (VALUES (1), (4)) s(a)" +
+                "   ON t.a = s.a) ts " +
+                "FULL OUTER JOIN (VALUES (2), (5)) r(a) " +
+                "ON ts.a = r.a",
+                anyTree(
+                        project(
+                                ImmutableMap.of("expr", expression("coalesce(r, ts)")),
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java]: why `r` is on the left side, I mean why `coalesce(r, ts)` when predicate is `ON ts.a = r.a`?

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestFullOuterJoinWithCoalesce.java]: For full other join, if the input is partitioned on both `ts.a` and `r.a`, the output is partitioned on both `coalesce(ts.a, r.a)` and `coalesce(r.a, ts.a)`. The order of the coalesce should not matter.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java [hunk 2]
```diff
@@ -275,13 +311,19 @@ public boolean isConstant()
 
         public boolean isVariable()
```
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: It doesn't make sense for the same thing being referenced interchangeably with `variable`, `column` and `symbol` to me. But that's existing code so I'm just ignoring it. I have no objection in changing them all to `symbol`. Happy to do that as a separate commit if you like.

[the remark this answers is not in the document]
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: I don't have a strong opinion here. I'm leaving it up to you.

**Comment by wenleix** [comment]
Curious: What's the change made with the revert one? -- I could manually do diff though ... ;)

**Comment by rongrong** [comment]
Addressed @martint's comment at https://github.com/prestosql/presto/pull/262#discussion_r258687795

@wenleix PrestoSQL has already done that for you. ;)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java]: Added a diff to do the renaming.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java [hunk 3]
```diff
@@ -833,4 +849,35 @@ else if (equality.getRight().equals(column) && columns.contains(equality.getLeft
...
+                List<Expression> assignmentOperands = ((CoalesceExpression) entry.getValue()).getOperands();
+                if (!assignmentOperands.stream().allMatch(SymbolReference.class::isInstance)) {
+                    return Optional.empty();
+                }
+
+                Set<Symbol> symbolsInAssignment = assignmentOperands.stream()
+                        .filter(SymbolReference.class::isInstance)
+                        .map(Symbol::from)
+                        .collect(toImmutableSet());
+                if (symbolsInAssignment.equals(symboslInExpression)) {
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: Is this comment addressed? https://github.com/prestosql/presto/pull/262/files#r258687795

Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: Can we just do 
```java
  if (symbolsInAssignment.size() == assignmentOperands.size() && symbolsInAssignment.equals(symboslInExpression)) {
...
}
```

And replace the condition on line 868? -- I could be wrong about their semantic equivalence.

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: They would effectively have the same results, but they don't have the same semantic meaning.

**Comment by rongrong** [comment]
Incorporated some of the changes from https://github.com/prestosql/presto/pull/288 as well

**Pull request closed without merging**